server/models/UserTwoFactorMethod.ts

```diff
@@ -0,0 +1,101 @@
+import {
+  BelongsToGetAssociationMixin,
+  CreationOptional,
+  DataTypes,
+  ForeignKey,
+  InferAttributes,
+  InferCreationAttributes,
+  Model,
+} from 'sequelize';
+
+import sequelize from '../lib/sequelize';
+import { TwoFactorMethod } from '../lib/two-factor-authentication';
+
+import User from './User';
+
+export type UserTwoFactorMethodTOTPData = {
+  secret: string;
+};
+
+export type UseTwoFactorMethodYubikeyOTPData = {
+  yubikeyDeviceId: string;
+};
+
+export type UserTwoFactorMethodData = {
+  [TwoFactorMethod.TOTP]: UserTwoFactorMethodTOTPData;
+  [TwoFactorMethod.YUBIKEY_OTP]: UseTwoFactorMethodYubikeyOTPData;
+};
+
+export default class UserTwoFactorMethod<
+  T extends Exclude<TwoFactorMethod, TwoFactorMethod.RECOVERY_CODE>,
+> extends Model<InferAttributes<UserTwoFactorMethod<T>>, InferCreationAttributes<UserTwoFactorMethod<T>>> {
+  declare id: CreationOptional<number>;
+  declare method: T;
+
+  declare UserId: ForeignKey<User['id']>;
+  declare User?: User;
+  declare getUser: BelongsToGetAssociationMixin<User>;
+
+  declare data: CreationOptional<UserTwoFactorMethodData[T]>;
+
+  declare createdAt: CreationOptional<Date>;
+  declare updatedAt: CreationOptional<Date>;
+  declare deletedAt: CreationOptional<Date>;
+
+  static async userMethods(userId: number): Promise<TwoFactorMethod[]> {
+    const result = await UserTwoFactorMethod.findAll({
+      attributes: ['method'],
+      group: 'method',
+      where: {
+        UserId: userId,
+      },
+    });
+
+    return result.map(r => r.method);
+  }
+}
+
+UserTwoFactorMethod.init(
+  {
+    id: {
+      allowNull: false,
+      autoIncrement: true,
+      primaryKey: true,
+      type: DataTypes.INTEGER,
+    },
+    method: {
+      type: DataTypes.STRING,
+      allowNull: false,
+    },
+    data: {
+      type: DataTypes.JSONB,
+      allowNull: true,
+    },
```

Comment:
Can we count on TS to statically validate the data content for us here or should we also validate the structure against what is expected from the method?

Reply:
If you are interacting with this model from a typescript file, you will get a static check which is very convenient already. I used a generic type so you get a correct validation for the appropriate two factor method. This is good but bypassed as you know.

I would love to get a runtime schema validation on the data fields in general and I think we can experiment with something here if its simple enough.

Reply:
@kewitz I added a runtime validator with [zod](https://github.com/colinhacks/zod), it ties up nicely with typescript.